source/AI.cpp

```diff
@@ -621,19 +647,38 @@ void AI::Step(const PlayerInfo &player, Command &activeCommands)
 
 		// Pick a target and automatically fire weapons.
 		shared_ptr<Ship> target = it->GetTargetShip();
+		// Find a targeted asteroid but only if the ship is player-owned.
+		shared_ptr<Minable> targetAsteroid = FindTargetAsteroid(*it);
 		if(isPresent && !personality.IsSwarming())
 		{
 			// Each ship only switches targets twice a second, so that it can
 			// focus on damaging one particular ship.
 			targetTurn = (targetTurn + 1) & 31;
-			if(targetTurn == step || !target || target->IsDestroyed() || (target->IsDisabled()
-					&& personality.Disables()) || !target->IsTargetable())
+			if(targetTurn == step && targetAsteroid)
+				it->SetTargetAsteroid(targetAsteroid);
+			else if(targetTurn == step || !target || target->IsDestroyed()
+				|| (target->IsDisabled() && personality.Disables()) || !target->IsTargetable())
 				it->SetTargetShip(FindTarget(*it));
 		}
 		if(isPresent)
 		{
-			AimTurrets(*it, firingCommands, it->IsYours() ? opportunisticEscorts : personality.IsOpportunistic());
-			AutoFire(*it, firingCommands);
+			// Player-owned ships should harvest asteroids when they are destroyed.
+			if(HarvestAfterAsteroidMining(*it) && DoHarvesting(*it, command))
```

Comment:
Is there a particular reason to do this check here, rather than integrating with the existing mining + harvesting code, or e.g. FollowOrders?

Reply:
I am new to the code base and not sure I understand.  I put it here because that's where I was able to get it to work.  I tried several iterations of getting the harvest behavior to work and found difficulty in finding the right spot.

If you could guide me where it would be most appropriate I'd be happy to make the change.

Reply:
I think `HarvestAfterAsteroidMining` is necessary because of the transition of `Orders::MINING` -> `Orders::HARVEST`.  I tried to reuse most of the AI code and `DoHarvesting` is itself the same that the NPC AI uses.

Reply:
> Mostly I hate introducing additional special-casing in the core logic when we have the personality and orders system for handling behaviors and commands.

Considering I'm a bit new to the code base I'm not sure I understand what change you'd like me to make with this.

> Furthermore, at no time should player escorts prioritize an asteroid over a hostile ship, _even if they have a harvest or mine order_. (We have a responsibility to not make it easy for a player to get their escorts killed.)

This position is non-intuitive for a player.  If I push F to fight my target.  I know _for fact_ that the target will be fought by my fleet.  If that target is a friendly... the fleet will ignore hostiles and attack the friendly.  If the target is a hostile, the fleet ignores ships killing them to go after the hostile (even to their own detriment).

If we make mining behave differently... i.e. "fight my target but not really because if a hostile comes in the system don't fight my target", this will perplex the player... because the target they want fought is being ignored.  I think this would be reported as a bug and I do not want it to behave this way.

It is _the player's responsibility_ to ensure their escorts do not get killed.  If they choose to focus their fleet on a ship faster than their escorts and enemies (like interceptors) shred their escorts while they're trying to chase... then that's a lesson in fleet combat for the player to pay more attention to the battle.  By extension if they're chasing an asteroid and a fleet comes along and melts their escort mining that lesson is in the same light as focused combat.

> Based on what I see, if the player is mining in a hostile system, they'll be using the "target asteroid" functionality (lest their active target become the nearest hostile), which when paired with this will cause their fleet to try to harvest the flotsam rather than disable hostiles.

When I am playing testing this pull request; the fleet does prioritize hostiles over flotsam.  They don't try to harvest above fighting.  Can you explain more why you think the logic renders that to be the case here?

> It's fine for them to harvest the flotsam, but not fine for them to stop attacking and engaging enemy ships to do so when it is an "implicit" order. (Compare with move / hold / gather, which are explicitly issued orders.)

V toggles harvest on and off.  It is not intended to be an implicit order.  When a player presses V to target a message appears `fleet is preparing to harvest` letting the player know they're in harvest mode now.  When they push F to attack the asteroid the asteroid breaks up and the fleet harvests.  This is a pretty intuitive behavior for a player.  They don't know under the hood there's a transition.  From a play-ability perspective the order is explicit.

Multiple players play testing have said this is the kind of behavior they'd like to see.  Originally when I opened this pull request it did not have the feature.  It was added later after multiple people play testing requested it (and I also felt the gap when I was playing).

If the player presses V multiple times with the same asteroid targeted or no asteroid targeted the player gets toggle messages like `fleet is preparing to harvest` and `fleet is no longer preparing to harvest`.

> Moving this handling to the "FollowOrders" method will help respect the existing escort behavior hierarchy.

I can move it but I'd want to to behave the same way it does now as it is being played.  It is pretty intuitive the way it works like other "fight my target" commands.

Reply:
> Moving this handling to the "FollowOrders" method will help respect the existing escort behavior hierarchy.

Upon closer inspection.  The important part here for harvesting is `continue`.  I don't think I can move this code into FollowOrders.

Reply:
FollowOrders is called too late; however I've figured out a way to overcome this.  I'm going to push a commit now.